### PR TITLE
ERC20 Arb take

### PR DESCRIPTION
**Pull request by grandizzy**
- changed pool's `arbTake` signature and removed `amount` parameter (arb is bounded by amount of borrower debt / collateral and bucket deposit) 
- added `AuctionPriceGteQArbPrice` error to revert when price of arbed bucket is lower than auction price
- included `isReward` flag in `ArbTake` event
- added `Auctions._validateTake` that is called by both `*take` methods (and will probably be called by `depositTake` too): check if auction is valid to be taken and calculates values for take action
- added `Buckets.addLPs` to add LP amount in bucket and lender's accumulators

Tests:
- moved all `*take` assertion from `ERC20DSTestPlus` to `DSTestPlus`
- `testArbTakeLTNeutralPrice` and `testArbTakeGTNeutralPrice` + revert tests `testArbTakeReverts`. Other tests pending. TODO: in a different PR we should split auction related tests in different contracts as `ERC20PoolLiquidationsTest` counts now 2160 LOCs and it becomes hard to follow / maintain
- added gas load test for `arbTake` with different way of deposits setup

gas results (`arbTake` test skips hours to account interest rate accumulation while regular `take` doesn't hence the difference in gas)
```
│ addQuoteToken                              ┆ 91205           ┆ 153360 ┆ 132790 ┆ 638279  ┆ 47999   │
│ arbTake                                    ┆ 400167          ┆ 400191 ┆ 400191 ┆ 400215  ┆ 2       │
│ borrow                                     ┆ 167250          ┆ 226820 ┆ 192581 ┆ 768700  ┆ 104002  │
│ kick                                       ┆ 173453          ┆ 252313 ┆ 246817 ┆ 1078412 ┆ 32000   │
│ moveQuoteToken                             ┆ 269533          ┆ 269533 ┆ 269533 ┆ 269533  ┆ 1       │
│ pledgeCollateral                           ┆ 62635           ┆ 65226  ┆ 64870  ┆ 508922  ┆ 96001   │
│ removeAllQuoteToken                        ┆ 149391          ┆ 151566 ┆ 151274 ┆ 165987  ┆ 2000    │
│ removeQuoteToken                           ┆ 184168          ┆ 189999 ┆ 189793 ┆ 207041  ┆ 2000    │
│ repay                                      ┆ 483292          ┆ 558028 ┆ 569947 ┆ 809526  ┆ 16001   │
│ take                                       ┆ 51484           ┆ 55208  ┆ 54672  ┆ 274626  ┆ 15998   │
```
compared with `develop`
```
│ addQuoteToken                              ┆ 91148           ┆ 156506 ┆ 133879 ┆ 638222  ┆ 43999   │
│ borrow                                     ┆ 167250          ┆ 231927 ┆ 191892 ┆ 768700  ┆ 88002   │
│ kick                                       ┆ 173475          ┆ 249007 ┆ 244636 ┆ 1078434 ┆ 16000   │
│ moveQuoteToken                             ┆ 288482          ┆ 288482 ┆ 288482 ┆ 288482  ┆ 1       │
│ pledgeCollateral                           ┆ 62657           ┆ 64767  ┆ 64836  ┆ 508944  ┆ 80001   │
│ removeAllQuoteToken                        ┆ 149413          ┆ 151586 ┆ 151292 ┆ 168363  ┆ 2000    │
│ removeQuoteToken                           ┆ 184190          ┆ 190022 ┆ 189815 ┆ 209403  ┆ 2000    │
│ repay                                      ┆ 483292          ┆ 558028 ┆ 569947 ┆ 809526  ┆ 16001   │
│ take                                       ┆ 51387           ┆ 55130  ┆ 54594  ┆ 274548  ┆ 15998   │
```

### CODE REVIEW

File: src/erc20/ERC20Pool.sol [hunk 1]
```diff
@@ -147,9 +147,43 @@ contract ERC20Pool is IERC20Pool, Pool {
...
+            uint256 quoteTokenAmount,
+            uint256 collateralArbed,
+            uint256 auctionPrice,
+            uint256 bondChange,
+            bool isRewarded
+        ) = auctions.arbTake(liquidation, borrower, bucketDeposit, poolState.inflator);
+
+        // cannot arb with a price lower than or equal with the auction price
+        uint256 bucketPrice = PoolUtils.indexToPrice(index_);
+        if (auctionPrice >= bucketPrice) revert BadArbTakePrice();
```
Comment (ith-harvey) [inline src/erc20/ERC20Pool.sol]: Would rename revert to `revert AuctionPriceGTEQArbPrice()`

File: src/erc20/ERC20Pool.sol [hunk 2]
```diff
@@ -147,9 +147,43 @@ contract ERC20Pool is IERC20Pool, Pool {
...
+
+        // collateral is moved from the loan to the bucket’s claimable collateral
+        borrower.collateral         -= collateralArbed;
+        buckets[index_].collateral += collateralArbed;
```
Comment (ith-harvey) [inline src/erc20/ERC20Pool.sol]: nit: formatting assignment

File: src/libraries/Auctions.sol [hunk 3]
```diff
@@ -231,6 +231,57 @@ library Auctions {
...
+        if (borrowerDebt >= bucketDeposit_) {
+            collateralArbed_ = Maths.wdiv(bucketDeposit_, Maths.wmul(auctionPrice_, factor));
+        } else {
+            collateralArbed_ = Maths.wdiv(borrowerDebt, Maths.wmul(auctionPrice_, factor));
+        }
+
+        if (collateralArbed_ > borrower_.collateral) collateralArbed_ = borrower_.collateral;
+        quoteTokenAmount_ = Maths.wmul(Maths.wmul(factor, collateralArbed_), auctionPrice_);
+
+        if (!isRewarded_) {
```
Comment (ith-harvey) [inline src/libraries/Auctions.sol:283]: @mattcushman can you confirm... is there a possible case where the BPF is negative and the kicker is penalized?

Since the `auctionPrice` is always less then the `HPB` my understanding is that they kicker is never penalized and always rewarded.

File: src/libraries/Auctions.sol [hunk 4]
```diff
@@ -231,6 +231,57 @@ library Auctions {
...
+        uint256 bondChange_,
+        bool    isRewarded_
+    ) {
+        uint256 borrowerDebt;
+        int256  bpf;
+        uint256 factor;
+        (auctionPrice_, borrowerDebt, bpf, factor, isRewarded_) = _validateTake(liquidation_, borrower_, poolInflator_);
+
+        // determine how much of the loan will be repaid
+        if (borrowerDebt >= bucketDeposit_) {
```
Comment (ith-harvey) [inline src/libraries/Auctions.sol:267]: Should we perform a check `bucketDeposit_ != 0 && borrowerDebt != 0 && borrowerCollateral != 0` ? (remember an auction could be active have debt  but no collateral... as we have seen in tests : ))

Also I think this code could be simplified to:
```
uint256 depositToArb = Maths.min(bucketDeposit_, borrowerDebt);
collateralArbed_         = Maths.wdiv(depositToArb, Maths.wmul(auctionPrice_, factor))
```

File: src/libraries/Auctions.sol (hunk 3 above)
Comment (ith-harvey) [reply src/libraries/Auctions.sol:283]: Actually now that I think of it yes this situation could happen. A lender could deposit after the loan has been kicked but before the debt has been taken... this situation would penalize the kicker.

**Review by ith-harvey — COMMENTED** [review]
Its a WIP let's keep chipping

File: src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol [hunk 5]
```diff
@@ -1653,18 +1653,293 @@ contract ERC20PoolLiquidationsTest is ERC20HelperContract {
...
+                borrowerDebt:              19.778761259189860403 * 1e18,
+                borrowerCollateral:        2 * 1e18,
+                borrowerMompFactor:        9.917184843435912074 * 1e18,
+                borrowerCollateralization: 0.983003509435146965 * 1e18
+            }
+        );
+
+        // --- GET OVERFLOW HERE IF I UNCOMMENT ARBTAKE--
+
+        // _arbTake(
```
Comment (ith-harvey) [inline src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol]: I get an over/underflow when I uncomment the `arbTake` here -> https://github.com/ajna-finance/contracts/blob/21d2e82fe5d27370d69c99d22c52b26ea3bd229c/src/base/Pool.sol#L513

This is due to the arb take attempting to repay more debt then exists in the loan. I think this issue arrises here inside `auctions.arbTake` -> https://github.com/ajna-finance/contracts/blob/21d2e82fe5d27370d69c99d22c52b26ea3bd229c/src/libraries/Auctions.sol#L272

As that is where we determine the quoteTokenAmount to be repaid although I'm still not sure why that amount is greater then borrower debt...

File: src/libraries/Auctions.sol (hunk 3 above)
Comment (ith-harvey) [reply src/libraries/Auctions.sol:283]: added that here but it overflows -> https://github.com/ajna-finance/contracts/pull/312/commits/20647e72eef345cfb282d71ed0d530d17a23798e

File: src/libraries/Auctions.sol (hunk 4 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol:267]: Check for collateral is done in ERC20Pool.arbTake

I wrote it this way to avoid stack too deep, I also think that's better readable, as Matt pointed out in some other reviews.

File: src/libraries/Auctions.sol (hunk 3 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol:283]: most probably due to the TODO here https://github.com/ajna-finance/contracts/pull/312/files#diff-f60f760fc956e7ea35a0c59d0aee29a65646591dc54271fff50ddc44c06ce7f3R184
`TODO should quoteTokenAmount be expressed in t0 debt or it is already?`

Comment (grandizzy) [reply src/libraries/Auctions.sol:283]: fixed with https://github.com/ajna-finance/contracts/pull/312/commits/24700909fb732c9a952e46738c00fa13c055282e

File: src/erc20/ERC20Pool.sol (hunk 1 above)
Comment (grandizzy) [reply src/erc20/ERC20Pool.sol]: changed in https://github.com/ajna-finance/contracts/pull/312/commits/5dcb833b7dc3e152e31a35b974897eed69453444

File: src/erc20/ERC20Pool.sol (hunk 2 above)
Comment (grandizzy) [reply src/erc20/ERC20Pool.sol]: changed in https://github.com/ajna-finance/contracts/pull/312/commits/10a90e0b58ba54917c77200b2fa98308019ef4f8

File: src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol (hunk 5 above)
Comment (grandizzy) [reply src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol]: fixed with https://github.com/ajna-finance/contracts/commit/24700909fb732c9a952e46738c00fa13c055282e

File: src/libraries/Auctions.sol (hunk 4 above)
Comment (ith-harvey) [reply src/libraries/Auctions.sol:267]: Sorry George, is that in response to this? -> 
```
Should we perform a check bucketDeposit_ != 0 && borrowerDebt != 0 && borrowerCollateral != 0 ? (remember an auction could be active have debt but no collateral... as we have seen in tests : ))
```

Comment (grandizzy) [reply src/libraries/Auctions.sol:267]: yeah, Check for collateral is done in `ERC20Pool.arbTake`

File: src/libraries/Auctions.sol [hunk 6]
```diff
@@ -231,6 +231,66 @@ library Auctions {
...
+            t0repayAmount_    = borrower_.t0debt;
+            quoteTokenAmount_ = Maths.wmul(t0repayAmount_, poolInflator_);
+        }
+
+        collateralArbed_ = Maths.wdiv(quoteTokenAmount_, auctionPrice_);
+
+        if (collateralArbed_ > borrower_.collateral) {
+            collateralArbed_  = borrower_.collateral;
+            quoteTokenAmount_ = Maths.wmul(collateralArbed_, auctionPrice_);
+            t0repayAmount_    = Maths.wdiv(quoteTokenAmount_, poolInflator_);
```
Comment (mattcushman) [inline src/libraries/Auctions.sol]: These should be:
t0repayAmount_ = Maths.wdiv(Maths.wmul(factor,quoteTokenAmount_), poolInflator_);

File: src/libraries/Auctions.sol [hunk 7]
```diff
@@ -231,6 +231,66 @@ library Auctions {
...
+    ) {
+        uint256 borrowerDebt;
+        int256  bpf;
+        uint256 factor;
+        (auctionPrice_, borrowerDebt, bpf, factor, isRewarded_) = _validateTake(liquidation_, borrower_, poolInflator_);
+
+        // determine how much of the loan will be repaid
+        if (borrowerDebt >= bucketDeposit_) {
+            t0repayAmount_    = Maths.wdiv(bucketDeposit_, poolInflator_);
+            quoteTokenAmount_ = bucketDeposit_;
```
Comment (mattcushman) [inline src/libraries/Auctions.sol]: I think this should be Maths.wdiv(bucketDeposit_, factor);

File: src/libraries/Auctions.sol [hunk 8]
```diff
@@ -231,6 +231,66 @@ library Auctions {
...
+        } else {
+            t0repayAmount_    = borrower_.t0debt;
+            quoteTokenAmount_ = Maths.wmul(t0repayAmount_, poolInflator_);
```
Comment (mattcushman) [inline src/libraries/Auctions.sol]: Should be 
quoteTokenAmount_ = Maths.wdiv(Maths.wmul(t0repayAmount_, poolInflator_), factor);;

File: src/erc20/ERC20Pool.sol [hunk 9]
```diff
@@ -147,9 +147,68 @@ contract ERC20Pool is IERC20Pool, Pool {
...
+                    liquidation.kicker,
+                    Maths.wrdivr(bondChange, bucketExchangeRate));
+            }
+
+            // collateral is moved to the bucket’s claimable collateral
+            bucket.collateral += collateralArbed;
+        }
+
+        // quote token are removed from the bucket’s deposit
+        deposits.remove(index_, quoteTokenAmount);
```
Comment (mattcushman) [inline src/erc20/ERC20Pool.sol]: This should be quoteTokenAmount*factor.   Equivalently (and easier to implement and perhaps more intuitive) here would be to use:

if (isRewarded)
  deposits.remove(index_, quoteTokenAmount+bondChange);
else
  deposits.remove(index_, quoteTokenAmount);

**Review by mattcushman — CHANGES_REQUESTED** [review]

File: src/erc20/ERC20Pool.sol (hunk 9 above)
Comment (grandizzy) [reply src/erc20/ERC20Pool.sol]: added in https://github.com/ajna-finance/contracts/pull/312/commits/7ca0040cc2a23c2e0bffe63d9d367e5fb5e42957

File: src/libraries/Auctions.sol (hunk 8 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol]: changed in https://github.com/ajna-finance/contracts/pull/312/commits/7ca0040cc2a23c2e0bffe63d9d367e5fb5e42957

File: src/libraries/Auctions.sol (hunk 7 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol]: changed in https://github.com/ajna-finance/contracts/pull/312/commits/7ca0040cc2a23c2e0bffe63d9d367e5fb5e42957

File: src/libraries/Auctions.sol (hunk 6 above)
Comment (grandizzy) [reply src/libraries/Auctions.sol]: changed in https://github.com/ajna-finance/contracts/pull/312/commits/7ca0040cc2a23c2e0bffe63d9d367e5fb5e42957

**Review by EdNoepel — COMMENTED** [review]
Quite happy with the `_validateTake` and `addLPs` refactoring.
Agree about splitting up liquidation tests into several files/contracts in another PR.  Happy to help with that after this is merged.

**Comment by grandizzy** [comment]
> Quite happy with the `_validateTake` and `addLPs` refactoring. Agree about splitting up liquidation tests into several files/contracts in another PR. Happy to help with that after this is merged.

Sounds great!
Re refactoring `addLPs` - I am still thinking if bucket accumulator should also be updated here (as it is PR) but rather rename `addLPs` to `addLenderLPs` and only update lenders balance... reason for this is that if we have to update balances for 2 lenders (as in arbTake case for kicker and taker) there will be 2 storage writes to update bucket lps... (whereas if updated outside of the function there will be only one, but probably increase contract size). What's your take on this?

File: src/erc20/ERC20Pool.sol [hunk 10]
```diff
@@ -147,9 +147,71 @@ contract ERC20Pool is IERC20Pool, Pool {
     /*** Pool External Functions ***/
     /*******************************/
 
-    function arbTake(address borrower_, uint256 amount_, uint256 index_) external override {
-        // TODO: implement
-        emit ArbTake(borrower_, index_, amount_, 0, 0);
+    function arbTake(
```
Comment (mattcushman) [inline src/erc20/ERC20Pool.sol:150]: What is there in this that is specific to erc20 tokens as opposed to NFTs?  In theory arbTake should be identical between the two, including allowing arbitrary fraction quantities of NFT to be taken.

File: src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol [hunk 11]
```diff
@@ -1777,18 +1777,383 @@ contract ERC20PoolLiquidationsTest is ERC20HelperContract {
...
+            {
+                index:        _i9_91,
+                lpBalance:    2_000 * 1e27,
+                collateral:   0,
+                deposit:      2_027.006589100074751447 * 1e18,
+                exchangeRate: 1.013503294550037375723500000 * 1e27
+            }
+        );
+
+        _arbTake(
```
Comment (mattcushman) [inline src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol:1911]: I think it would be a good idea to add an assertion for the pool reserves before and after takes as well.  The reserves should remain invariant.

File: src/erc20/ERC20Pool.sol (hunk 10 above)
Comment (grandizzy) [reply src/erc20/ERC20Pool.sol:150]: Unfortunately at this moment ERC721 pool contract size exceeds the limit if `arbTake` logic is moved in base class, will keep it for ERC20 pool only for now and check to make some more space for NFT in a follow up PR

File: src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol (hunk 11 above)
Comment (grandizzy) [reply src/_test/ERC20Pool/ERC20PoolLiquidations.t.sol:1911]: assert it for both `take` and `arbTake` here https://github.com/ajna-finance/contracts/pull/312/commits/c29ded94ba9577f5bdefe5b5c637de980a61e374

**Review by mattcushman — COMMENTED** [review]
lgtm

**Review by mattcushman — APPROVED** [review]
